### PR TITLE
IEP-683 Updating esp-idf master directly from plugin

### PR DESCRIPTION
**Pull request by sigmaaa**
## Description

Please include a summary of the change and which issue is fixed.

Fixes # ([IEP-683](https://jira.espressif.com:8443/browse/IEP-683))

## Type of change

- New feature (non-breaking change which adds functionality)

## How has this been tested?

Case 1:
- Download and select the ESP-IDF master version (Espressif -> Download and Configure ESP-IDF)
- By using a terminal reset the master version to the older commit (`git reset --hard <commit_id>`) and use `git clean -fd` to prevent untracked changes
- Update ESP-IDF from the plugin (Espressif -> ESP-IDF Tools Manager - Update ESP-IDF master)

Case 2:
-  Download and select the ESP-IDF release version (Espressif -> Download and Configure ESP-IDF)
- Check if the Update ESP-IDF command is not visible (Espressif -> ESP-IDF Tools Manager - Update ESP-IDF master)

**Test Configuration**:
* ESP-IDF Version:master
* OS (Windows,Linux and macOS): All

## Dependent components impacted by this PR:

- Espressif menu commands 

## Checklist
- [x] PR Self Reviewed
- [x] Applied Code formatting
- [x] Added Documentation
- [ ] Added Unit Test
- [x] Verified on all platforms - Windows,Linux and macOS

### CODE REVIEW

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/handlers/UpdateEspIdfHandler.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+	private List<String> getSubmoduleUpdateCommand()
+	{
+		List<String> command = new ArrayList<>();
+		command.add(gitExecutablePath);
+		command.add("submodule"); //$NON-NLS-1$
+		command.add("update"); //$NON-NLS-1$
+		command.add("--init"); //$NON-NLS-1$
+		command.add("--recursive"); //$NON-NLS-1$
+		return command;
+	}
```
Comment (alirana01) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/handlers/UpdateEspIdfHandler.java:108]: I think eclipse's own [org.eclipse.jgit.api.Git](https://download.eclipse.org/jgit/site/6.2.0.202206071550-r/apidocs/org/eclipse/jgit/api/package-summary.html) offers some better ways to do what we are trying to do manually here. You also can implement a progress monitor for that and it also provides a better way to work with the reporting on console. See if you can find anything in GitRepositoryBuilder.java class in our project in method public void repositoryClone(), although that particular implementation only takes care of cloning but seeing a few things in the documentation above you can see some methods that offer us other functionality like Pull. See this https://download.eclipse.org/jgit/site/6.2.0.202206071550-r/apidocs/org/eclipse/jgit/api/PullCommand.html

**Review by alirana01 — CHANGES_REQUESTED** [review]

**Comment by kolipakakondal** [comment]
Hi @sigmaaa Could you apply PR template for this so that it would help andrii to do the functional testing

File: bundles/com.espressif.idf.ui/OSGI-INF/l10n/bundle.properties [hunk 2]
```diff
@@ -54,4 +53,4 @@ command.label.10 = ESP-IDF: Python Clean
 command.tooltip.1 = Advised to run this after switching between IDF and Python versions.Generated python byte code can be deleted from the IDF directory
 command.name.12 = ESP-IDF: Python Clean
 command.name.11 = Delete
-
+command.name.14 = Update ESP-IDF master
```
Comment (alirana01) [inline bundles/com.espressif.idf.ui/OSGI-INF/l10n/bundle.properties:56]: Can you please name the property to reflect what it is referencing to meaningfully?

File: bundles/com.espressif.idf.ui/plugin.xml [hunk 3]
```diff
@@ -98,6 +98,11 @@
             id="com.espressif.idf.ui.command.eclipeguide"
             name="%command.name.9">
       </command>
+      <command
+            defaultHandler="com.espressif.idf.ui.handlers.UpdateEspIdfHandler"
+            id="com.espressif.idf.ui.command.update"
+            name="%command.name.14">
```
Comment (alirana01) [inline bundles/com.espressif.idf.ui/plugin.xml:104]: Can you please name the property to reflect what it is referencing to meaningfully?

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/UpdateEspIdfMasterPropertyTester.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import com.espressif.idf.core.logging.Logger;
+import com.espressif.idf.core.util.IDFUtil;
+
+public class UpdateEspIdfMasterPropertyTester extends PropertyTester
+{
+	@Override
+	public boolean test(Object receiver, String property, Object[] args, Object expectedValue)
+	{
+		// Check if the master version of ESP-IDF is being used. If false, do not show update command
+		String branchName = ""; //$NON-NLS-1$
```
Comment (alirana01) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/UpdateEspIdfMasterPropertyTester.java:22]: We can use StringUtil.EMPTY here for initializing empty strings

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/UpdateEspIdfMasterPropertyTester.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+		try
+		{
+			Git git = Git.open(new File(IDFUtil.getIDFPath()));
+			branchName = git.getRepository().getBranch();
+		}
+		catch (IOException e)
+		{
+			Logger.log(e);
+		}
+		return branchName.contentEquals("master"); //$NON-NLS-1$
```
Comment (alirana01) [inline bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/UpdateEspIdfMasterPropertyTester.java:32]: master can be moved to a private constant in the class, just a suggestion for consistent and clean style of code

File: bundles/com.espressif.idf.ui/OSGI-INF/l10n/bundle.properties (hunk 2 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.ui/OSGI-INF/l10n/bundle.properties]: Hi Ali, changed to command.name.UpdateEspIdfMaster. Thanks

File: bundles/com.espressif.idf.ui/plugin.xml (hunk 3 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.ui/plugin.xml]: Hi Ali, changed to the command.name.UpdateEspIdfMaster. Thanks

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/UpdateEspIdfMasterPropertyTester.java (hunk 4 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/UpdateEspIdfMasterPropertyTester.java]: added

File: bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/UpdateEspIdfMasterPropertyTester.java (hunk 5 above)
Comment (sigmaaa) [reply bundles/com.espressif.idf.ui/src/com/espressif/idf/ui/UpdateEspIdfMasterPropertyTester.java]: done. Thanks

**Comment by kolipakakondal** [comment]
Hi @sigmaaa 

I was already on master and while updating from eclipse I was having the below problem. I am on macOS.

Screenshot 2022-07-22 at 3 45 23 AM

```
Finished Checking out files 1, total 1 100% Completed
Checking out files 1, total 2 50% Completed
Checking out files 2, total 2 100% Completed
Finished Checking out files 2, total 2 100% Completed

```

**Comment by kolipakakondal** [comment]
Hi @sigmaaa Looks like you have added a couple of master merged PR here. Could you please clean that up?

**Comment by sigmaaa** [comment]
> Hi @sigmaaa Looks like you have added a couple of master merged PR here. Could you please clean that up?

done

**Review by alirana01 — APPROVED** [review]
LGTM

**Comment by sigmaaa** [comment]
Icon updated

**Comment by sigmaaa** [comment]
Updated README.md

**Review by kolipakakondal — APPROVED** [review]
LGTM

**Pull request merged**